Save latest post changes when exiting editor

mchowning opened the pull request:
Ensures that we have always saved the latest content from the editor when we exit the editor.

In addition, as [discussed in this comment](https://github.com/wordpress-mobile/WordPress-Android/pull/12144#issuecomment-645584192), in debug builds we are trying to make this issue extra visible by throwing an exception and crashing the editor, which causes the app to freeze for a few seconds and then recover outside the editor, adding a big red crash stacktrace to logcat. We obviously don't want to crash release builds, so instead we are sending an event to Sentry when there are changes to be saved because this means that the autosave mechanism failed. That event shows up looking like this: https://sentry.io/share/issue/f6224217e3e545cfb228982958317ef3/.

### Hiding the Whitespace Helps the Diff on This PR A Lot

When reviewing this, the diff is _much_ more manageable if you hide whitespace changes because there are a lot of indentation changes due to the callbacks wrapping a lot of code.

<img src="https://user-images.githubusercontent.com/4656348/90174574-38604700-dd74-11ea-81db-befedefe6de2.png" width="200" />

### To Test

Exit the editor before the 500ms debouncer has fired so there are "unsaved" changes.

1. Tap a character in the Gutenberg editor
2. Immediately press the back button to exit the editor
3. If a release build, an event similar to [this](https://sentry.io/share/issue/f6224217e3e545cfb228982958317ef3/) will be sent off to Sentry.  If a debug build, the editor will crash.
3. Confirm that the saved post includes the added character

PR submission checklist:

- [X] I have considered adding unit tests where possible.
- [X] I have considered adding accessibility improvements for my changes.
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

malinajirka commented:
Thanks @mchowning!

The current "architecture" of the EditPostActivity when it comes to saving is pretty simple. The Activity saves the post when it receives a "ping" from the editor that the content has changed (it adds 500ms delay/throttle). I personally really like this simplicity as it facilitates handling a lot of the edge case flows. The activity doesn't need to worry when it's supposed to save the post and when it isn't - it either received the ping or it didn't.
I think your solution works, however it feels like it adds extra complexity to the overall flow. I have been thinking about other possible solutions and I continued testing the apps and I came up with easy to reproduce steps.
1. Open a draft
2. Add "1" into the content and press the up button right after that
3. Open the post again and notice "1" is missing

<img src="https://user-images.githubusercontent.com/2261188/84235838-d3c80800-aaf6-11ea-8722-990c5054ebb0.gif" width="200px" />

I believe the issue is that the user is able to leave the editor during the "delay/throttle" we are adding to the save. So the post isn't saved even when the activity received the ping. How would you feel about [something like this](https://github.com/wordpress-mobile/WordPress-Android/compare/perform-pending-saves?expand=1). The code ensures that we perform the pending "save" before we exit the editor/publish the post. 

It doesn't handle the `When quickly entering text into a block (paragraph, verse, etc.), the autosave mechanism stops working.` scenario. However, I believe it's not the responsibility of the EditPostActivity to handle it. If we won't be able to fix this issue within gutenberg,  I guess we'll need to introduce a workaround in the EditPostActivity. However, I'd really like to consider that as a last resort as it make the EditPostActivity more complex. We've had 4 different implementations of the editor in the last 5 years. One thing we learned is that adding workarounds into the EditPostActivity for just one of the editors can bite us in the future.

Let me know what you think ;).

planarvoid commented:
hey @mchowning and @malinajirka 

> With respect to WPAndroid, the onSavePostTriggered event stream is used to kick off an asynchronous update without a callback, but immediately after we post to that event stream we check things like whether the post has changed or is publishable.

> You are absolutely right. Great catch! We'd need to wait for the result.

Do you mean that we might lose changes made in the past 500ms when we exit the editor too fast? If that's the case, we might want to implement a flag - something like `has pending saves` - and wait for it to finish before we leave the activity. This is quite an edge case though.

> If Gutenberg did change how often autosave pings were sent out it would be easy to miss this during testing

This should really be ensured on the Gutenberg part of code. We've implemented autosave to make sure the user never loses any changes. If we can't make sure Gutenberg triggers the `content change`, the user can easily achieve an inconsistent state or lose some work. We can't make sure the user always correctly finishes the editor (that's why we have autosave in the first place). You could make sure you trigger the `content change` event more often, for example when a user clicks on a block? Or loses focus on a block? Even if you're not sure there was a change. 
If we decide to implement saving on leaving the `EditPostActivity`, we should do it properly with a saving progress dialog. Retrieving the data and comparing if it has changed is a slow operation and that's why we made it asynchronous. We should make sure we don't make the UX any worse.

SergioEstevao commented:
For the record the way we achieved this on iOS is using a callback, when we press Save in the main apps, we ask Gutenberg for the latest version of the content, wait for the callback to return and the proceed with the save operation.

malinajirka commented:
@SergioEstevao Thanks for the info! What happens if the app crashes before the user hits "Save" or when the user leaves the app without hitting any button? I'm concerned that unless we automatically save all changes right after they happen we'll always have "content loss" issues. That's the main reason why Vojta and I decided for the mentioned approach which should handle all these scenarios. The worst thing that can happen is that the user loses changes from the last 500ms (or 50 last changes if they continuously keep typing).

SergioEstevao commented:
@malinajirka We needed to strike a balance between performance on the editor and the ability to keep data. 
Too much saves resulted in a jumping experience (because of too many requests for the full post HTML). 
We thought it was better to have better performance and make the editor more usable than been able to auto-save every keystroke.
In iOS, we auto-save every 5s, and I'm expecting that crashes will not be that frequent that the user will lose those 5s of typing that often.

malinajirka commented:
@SergioEstevao I completely agree that saving too often can have performance impact. That's one of the reasons why we are saving the content only after user makes some changes (with 500ms throttle). This delay/throttle is defined [here](https://github.com/wordpress-mobile/WordPress-Android/blob/20a93a60f1c0a89226a9a4de5b299fe4931342c5/WordPress/src/main/java/org/wordpress/android/ui/posts/editor/StorePostViewModel.kt#L31). I guess 500ms might not be enough for the user to press another key and I'm up for increasing it to 5s. (we also save the content after fifty changes to make sure we auto-save the content even when the user types continuously)  

However, it doesn't change my opinion about the approach taken in this PR. Imo the editor should still inform the app about all the keystrokes. The app can apply 5s delay instead of 500ms delay and save the post after that. 

We'd still need to implement [something like this](https://github.com/wordpress-mobile/WordPress-Android/compare/perform-pending-saves?expand=1) to perform the pending saves when the user is exiting the editor (it needs to be modified so the final save is performed synchronously - ideally with a progress dialog).

Wdyt?

mchowning commented:
Thanks for everyone's comments and feedback on this! It's been very helpful. :bow: 

@malinajirka and I chatted about this, and I think we've agreed on a way forward. At a high level, the plan is to:
* Implement a fail-safe save when exiting the editor to avoid any risk of users losing content
* During that save, check whether the saved post is different from the content reported by the editor. If they are different, send an exception to Sentry (and possibly have debug builds fail-fast by crashing) so that the failsafe save doesn't hide autosave breakages.
* Because the fail-safe save-on-exit will catch any changes that have not been processed because of the debouncer (currently changes that happen _just_ before exiting the editor can be lost), we do not expect that we will need to implement a separate fix for that.
* Nothing will change about how our autosave functionality **depends** on receiving "change pings" from the editor, so anything that causes the editor to not send those "change pings" to the main app is a bug that needs to be fixed in the editor.

I'm marking this PR as a draft while I work on updating this. When I'm done I'll reopen the PR for you all to look at and share your thoughts.

oguzkocer commented:
Hi there! `15.2` is being frozen, so this PR is being bumped to `15.3`. If these changes _have to_ be shipped with `15.2`, please target `release/15.2` and cc me in the PR so I can cut a new beta release. Also if you can keep the release dates in mind and bump the PRs & issues you are working on before the code freeze date, we'd really appreciate it.

mchowning committed:
Adding save on exiting the editor

mchowning commented:
:wave: @malinajirka, @planarvoid ! I finally got back to this and made the updates [discussed here](https://github.com/wordpress-mobile/WordPress-Android/pull/12144#issuecomment-645584192), and it's ready for another round of reviews. 

I have updated the PR description to reflect the latest changes.

malinajirka commented:
Thanks @mchowning! I was afk - I'll look into this tomorrow or on Thursday ;).

malinajirka reviewed (approved):
Thanks @mchowning! I've reviewed the code and tested the app and all works as described, great job! 

The only thing I'm not sure about is that we are aware of a bug (the 500ms debouncer issue) and we'll be crashing the app/logging events into sentry when it happens. I guess it might be interesting to see how many users actually encounter this issue in production, but it feels a bit weird to ignore the bug. Wdyt?

malinajirka commented:
I'm wondering if we could for example check if the debouncer is "in progress" and crash/log only when it isn't. Wdyt?

mchowning commented:
Thanks for the review @malinajirka ! 🙇 

> The only thing I'm not sure about is that we are aware of a bug (the 500ms debouncer issue) and we'll be crashing the app/logging events into sentry when it happens. I guess it might be interesting to see how many users actually encounter this issue in production, but it feels a bit weird to ignore the bug. Wdyt?

I kind of think of it not so much as we have a bug, but that we have a save mechanism that (a) includes an autosave mechanism that we know will not save changes made just before exiting the editor and (b) includes a fall-back save on exiting the editor that ensures there are no unsaved changes.

Like you said though, we are logging/crashing the app when what-I-just-claimed-wasn't-a-bug happens, which feels a bit odd.
 
> I'm wondering if we could for example check if the debouncer is "in progress" and crash/log only when it isn't. Wdyt?

My only concern with this is that it could hide some bugs in the autosave mechanism--for example if the debouncer somehow stopped ever triggering a save and instead stayed pending forever. If we continue logging _all_ the unsaved-changes-on-exit events, we would be able to catch the dramatic uptick in the number of those events if a bug like ^that^ was introduced.

With that said, it does seem _far_ more likely that the autosave mechanism gets broken by the editor's failure to send change events (as compared to the debouncer itself breaking), so realistically, even if we don't send events when the debouncer is "in progress", we will still likely send events for any bugs that come up. For that reason, I'm totally fine adding this check (I did a quick smoke test checking [saveJob](https://github.com/wordpress-mobile/WordPress-Android/blob/9049ef06c93c92af0d57f50f2b17b34bfd3cd828/WordPress/src/main/java/org/wordpress/android/ui/posts/editor/StorePostViewModel.kt#L44)`.isActive()` and it appeared to work well). Just let me know if you're still in favor of skipping the "event" if the debouncer is in progress, and I'll update the PR.

But maybe we can get the best of both worlds. We could send two different events to Sentry depending on whether the debouncer is "in progress". That way we would know that we expect the unsaved-changes-because-the-debouncer-in-progress event will regularly occur because of the known "bug", but we would also be able to check it and make sure it's remaining steady with no dramatic changes that would indicate a debouncer bug. At the same time, we would have the unsaved-changes-and-debouncer-is-NOT-in-progress event, and _any_ of those events would be evidence of a bug. wdyt?

malinajirka commented:
Thanks @mchowning!

>But maybe we can get the best of both worlds. We could send two different events to Sentry depending on whether the debouncer is "in progress". 

I like this idea. It should be rather simple to implement and we'll get the best of both worlds as you mentioned. My only minor concerns is that there are some limits on the number of events we can send to Sentry - but hopefully even the known bug should be quite rare.

------------

Btw I realized one issue with the implementation `mCrashLogging.report(message, T.EDITOR);` -> this method doesn't actually create an event in Sentry unless the app crashes afterwards. If I recall correctly Riccardo was testing these methods and the only method which creates an event is `reportException` (unless something has changed since than). We might want to verify that - let me know if you need a help with that.

mchowning committed:
Send different exceptions to Sentry based on autosave failure

mchowning commented:
I've updated the PR with sending separate events depending on whether the debouncer was pending @malinajirka .

> My only minor concerns is that there are some limits on the number of events we can send to Sentry - but hopefully even the known bug should be quite rare.

I wasn't aware of that limit--thanks for mentioning it. Like you said, this PR _shouldn't_ result in us sending too many events, so hopefully that isn't an issue here (and if it does send a lot of events, it's probably good that we find out because it means we need to figure out what is happening).

> Btw I realized one issue with the implementation mCrashLogging.report(message, T.EDITOR); -> this method doesn't actually create an event in Sentry unless the app crashes afterwards. If I recall correctly Riccardo was testing these methods and the only method which creates an event is reportException (unless something has changed since than).

That's interesting. I'm _pretty sure_ that I tested that without a crash occurring afterwards and saw the events in Sentry, but I can't swear to that with complete confidence. Regardless, with us now sending the separate events I wanted to create distinct exceptions to help make sure Sentry doesn't group them together, so I have switched to the `reportException` method, and I verified that it is working as expected in production builds ([example Sentry event](https://sentry.io/share/issue/9a96e3360d504bbb8d6a4b830558c7dd/)).

malinajirka reviewed (approved):
This is perfect! Thank you so much for working on all the changes 🙇!
